ssl/ssl_cert_comp.c

```diff
@@ -0,0 +1,484 @@
+        goto err;
+
+    ret = OSSL_COMP_CERT_helper(comp_data, comp_length, len, alg);
+    comp_data = NULL;
+
+ err:
+    OPENSSL_free(comp_data);
+    COMP_CTX_free(comp_ctx);
+    return ret;
+}
+
+void OSSL_COMP_CERT_free(OSSL_COMP_CERT *cc)
+{
+    int i;
+
+    if (cc == NULL)
+        return;
+
+    CRYPTO_DOWN_REF(&cc->references, &i, cc->lock);
+    REF_PRINT_COUNT("OSSL_COMP_CERT", cc);
+    if (i > 0)
+        return;
+    REF_ASSERT_ISNT(i < 0);
+
+    OPENSSL_free(cc->data);
+    CRYPTO_THREAD_lock_free(cc->lock);
+    OPENSSL_free(cc);
+}
+int OSSL_COMP_CERT_up_ref(OSSL_COMP_CERT *cc)
+{
+    int i;
+
+    if (CRYPTO_UP_REF(&cc->references, &i, cc->lock) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("OSSL_COMP_CERT", cc);
+    REF_ASSERT_ISNT(i < 2);
+    return ((i > 1) ? 1 : 0);
+}
+
+static int ssl_set_cert_comp_pref(int *prefs, int *algs, size_t len)
+{
+    int j = 0;
+    int i;
+    int found = 0;
+    int already_set[TLSEXT_comp_cert_limit];
+    int tmp_prefs[TLSEXT_comp_cert_limit];
+    int num_algs;
+
+    /* clear all algorithms */
+    if (len == 0 || algs == NULL) {
+        memset(prefs, 0, sizeof(tmp_prefs));
+        return 1;
+    }
+
+    num_algs = len / sizeof(*algs);
+
+    /* This will 0-terminate the array */
+    memset(tmp_prefs, 0, sizeof(tmp_prefs));
+    memset(already_set, 0, sizeof(already_set));
+    /* Include only those algorithms we support, ignoring duplicates and unknowns */
+    for (i = 0; i < num_algs; i++) {
+        switch (algs[i]) {
+        case TLSEXT_comp_cert_brotli:
+        case TLSEXT_comp_cert_zlib:
+        case TLSEXT_comp_cert_zstd:
+            /* Check for enablement */
+            if (!ossl_comp_has_alg(algs[i]))
+                break;
+            /* Check for duplicate */
+            if (already_set[algs[i]])
+                return 0;
+            tmp_prefs[j++] = algs[i];
+            already_set[algs[i]] = 1;
+            found = 1;
+            break;
+        default:
+            /* If it gets here, it's an unsupported algorithm and is ignored */
+            break;
+        }
+    }
+    if (found)
+        memcpy(prefs, tmp_prefs, sizeof(tmp_prefs));
+    return found;
+}
+
+static size_t ssl_get_cert_to_compress(SSL *ssl, CERT_PKEY *cpk, unsigned char **data)
+{
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(ssl);
+    WPACKET tmppkt;
+    BUF_MEM buf = { 0 };
+    size_t ret = 0;
+    const SSL_METHOD *method = NULL;
+
+    if (sc == NULL
+            || cpk == NULL
+            || !sc->server
+            || !SSL_in_before(ssl))
+        return 0;
+
+    /* Use the |tmppkt| for the to-be-compressed data */
+    if (!WPACKET_init(&tmppkt, &buf))
+        goto out;
+
+    /* no context present, add 0-length context */
+    if (!WPACKET_put_bytes_u8(&tmppkt, 0))
+        goto out;
+
+    /*
+     * ssl3_output_cert_chain() may generate an SSLfata() error,
+     * for this case, we want to ignore it
+     */
+    sc->statem.ignore_fatal = 1;
```

Comment:
Yuck. Can't we pass this as a param to `ssl3_output_cert_chain`?

Reply:
Yes, but It gets VERY ugly, considering how deep the call chain is.
I can throw this on as a separate fixup commit, so the two methods can be compared.

Reply:
I think I have something.